### ISSUE TITLE
Don't make gradle-wrapper.jar a default part of a new project

### ISSUE THREAD

**Issue by rosti-il** [gradle/gradle#25784]
### Expected Behavior

I expect any binary file and especially any executable Jar file be nether part of my projects nor saved in my VCS repositories. Having this in mind I want Gradle to stop putting the `gradle-wrapper.jar` file into the `gradle/wrapper` directory of any project and to add the `gradle/wrapper/*.jar` line into the `.gitignore` by default. I propose making this `gradle-wrapper.jar` downloadable and teach the `gradlew` and the `gradlew.bat` scripts to download the `gradle-wrapper.jar` file if missing, like any other dependency. Those scripts may try to use following ways to download the wrapper Jar file:

* Using an external tool, like `wget` or `curl`, if installed.
* Using a Java code that could be compiled and run or (with Java 11 and newer) just be run without compilation. That Java code could be putted into the `gradle/wrapper` directory by default.

I'm not against keeping the current behavior. I want just make it optional and not default.

### Current Behavior (optional)

Currently a command like `gradle init` creates a project with Gradle Wrapper, including the `gradle-wrapper.jar` file. Then this `gradle-wrapper.jar` file is usually added into the VCS repository of that project. For example the Gradle project itself also has it here in its GitHub repository.

Why this is a bad practice? Because:

* It is insecure by design. An executable or executed by a script Jar file can easily have a malware or any other unwanted behavior.
* Changes between versions of any binary file are unreadable or just not shown by VCS.
* Waste of the storage space in the VCS repository.

### Context

See the above.

**Comment by Techcable** [comment]
> Using a Java code that could be compiled and run or (with Java 11 and newer) just be run without compilation. That Java code could be putted into the gradle/wrapper directory by default

I think replacing the Gradle wrapper JAR with direct source-file execution is a really great idea!

We already require 

As of 2022, [Java 11](https://newrelic.com/resources/report/2022-state-of-java-ecosystem#toc-java-11-is-the-new-standard) has become more popular than Java 8 even in *production*. Java 8 is almost nine years old. More and more of the Java ecosystem is moving on and requiring Java 11. 

Given that Gradle already requires a JVM to run, it's not unreasonable to require an existing JDK 11+ installation for the wrapper.

This could all be configurable, but this is pretty much the sort of scripting/bootstrap situation (JEP 330)[https://openjdk.org/jeps/330] was designed for.

Requiring Java 11 could save too thanks to all the new convenience APIs that have been added. (One example is[`InputStream.transferTo(OutputStream)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/InputStream.html#transferTo(java.io.OutputStream)))

Optionally, another Python wrapper script could be offered to download the JVM if missing. Python is present on essentially all Unix machines and most Windows developer machines.

At least making these new ideas options would be nice, even if they aren't the default.

**Comment by ov7a** [comment]
Thank you for your interest in Gradle!

This feature request is in the backlog of the relevant team, but this area of Gradle is currently not a focus one, so it might take a while before it gets implemented.

---

Related:
* #2508

Some kind of bootstrapper script to download gradle wrapper and/or JDK might be useful to reduce VCS space/noise, but it requires careful design considering security and different platforms' support.

**Comment by rosti-il** [comment]
Hi @ov7a,

The [Allow Gradle to run without a JDK installed](https://github.com/gradle/gradle/issues/2508) proposal that you've pointed me out is not fully related to this issue. In contrast to this issue that proposal tries to make an architect level change that could allow one to use Gradle on systems without Java preinstalled and with isolation (containers). Not surprisingly that proposal is still open already for 6 years.

My proposal is about a completely different and simpler problem. It just tries to treat the `gradle-wrapper.jar` as any other dependency Jar in Java project. In case it's implemented in Java it will be multi-platform by design. Security related things could be treated similarly to how they are already treated in dependency downloading code. For example this is the official SHA256 hash of the wrapper that is already available: https://services.gradle.org/distributions/gradle-8.2.1-wrapper.jar.sha256 and Java has a built into its standard library capabilities to check that hash against the actually downloaded `gradle-wrapper.jar` file. This `gradle-wrapper.jar` file could also be deployed in Maven repository as any other dependency where it has `md5` and `sha1` hashes and optionally [from Gradle 6.0](https://github.com/gradle/gradle/pull/11053) `sha256` and `sha512`. All of these hashes could be checked by Java standard library. I don't see here any special technical problem that can prevent making it secure and multi-platform. Do you see any?

I think I can implement this by myself. I just wanted to hear opinions, especially from the Gradle project members that will then consider to approve and merge my pull request or not. This is why I started with this issue and not with a pull request.

**Comment by ov7a** [comment]
Yes, I'm aware that the issue I pointed out is loosely related, however, it is something that should be considered when work on your issue happens.

Thank you for providing more thoughts on this topic. I'm not ready to answer your question right now, sorry.

It's awesome that you want to work on this issue! However, this is quite a complex change in terms of security implications and user experience. As stated in our [contributing guide](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md#before-you-start), before you start implementing this, the implementation idea should be discussed with Gradle engineers responsible for that area. So, please provide more details for your ideas. For example, what should change for existing wrapper use cases? What should happen if there is a corporate proxy? Etc.

Please note that our capacity is limited, so the response time may vary. Thanks again for your help!

**Comment by ov7a** [comment]
@rosti-il You can use our spec template and fill it [here](https://docs.google.com/document/d/1oOhJrsLnY47BDSVNJGTiZ4dPkgfw2BjrcA6SW5zsn3U/edit?usp=sharing) (I don't know your email, please write me to vchesnokov [at] gradle.com so I can give you editing access)

**Comment by hfhbd** [comment]
I really like the JEP 330 idea which could be easily implemented by simply moving the content of the wrapper into one (big) Java class. While this will only work with a JVM 11+, this could be handled by a property in the wrapper task to either generate a jar + shell scripts or the Java class.

**Comment by jvandort** [comment]
One major concern is startup speed. Gradle makes a significant effort towards decreasing build times, and for small builds, the startup speed of the launcher cannot be ignored. Whatever changes we make to the launcher startup must be benchmarked. Bash, python, java, etc launchers all seem possible, though we would need to test speed of startup.

Though in all likelihood, we will need to have bash and batch implementations of the wrapper, since its always been like that, and changing it would be disruptive. For this reason, I would say a python wrapper script is probably not likely. 

The argument for a Java 11 code script is interesting though. Speed is the biggest factor here. But this is more reasonable than python, as we can likely expect Gradle users to have java installed (though, it would be nice to not have this requirement)

We have toyed around with the idea of a native launcher, for even faster startup speed, though that is likely some time away. However, if we went with this approach we would likely need to invest in developing a downloader for that launcher -- which would ideally be text-based. 

I think either a Java 11 implementation or bash/batch implementations are the most likely to be accepted. Requiring Java 11 to launch Gradlew will be a harder sell for sure though. 

I think we should leave the JDK downloading topic for another discussion, as I think its a separate issue.
